File: apps/vision-service/src/vision_service/events/envelope.py
"""Event envelope — Pydantic mirror of packages/event-schemas' TS types
(src/envelope.ts) and src/schemas/event-envelope.schema.json, per
docs/architecture/Coding_Standards.md's Events section and
docs/adr/ADR-005-event-schema-versioning.md's versioning policy.

Field names are deliberately camelCase (not idiomatic Python) so this
model, the JSON Schema, and the TS type share one vocabulary — that
parity is exactly what
apps/vision-service/tests/test_event_schema_sync.py checks in CI
(REQ-3.4). `# noqa: N815` per-file-ignore is set in pyproject.toml for
this reason.
"""

import uuid
from datetime import datetime, timezone
from typing import Generic, TypeVar

from pydantic import BaseModel, ConfigDict

TPayload = TypeVar("TPayload")

# Membership (not order) must match event-envelope.schema.json's
# `properties` keys and packages/event-schemas/src/envelope.ts's
# ENVELOPE_FIELD_NAMES exactly.
ENVELOPE_FIELD_NAMES = (
    "eventId",
    "eventType",
    "eventVersion",
    "occurredAt",
    "correlationId",
    "causationId",
    "producer",
    "payload",
)


class EventEnvelope(BaseModel, Generic[TPayload]):
    model_config = ConfigDict(frozen=True)

    eventId: str
    eventType: str
    eventVersion: int
    occurredAt: str
    correlationId: str
    causationId: str | None
    producer: str
    payload: TPayload


def create_envelope(
    *,
    event_type: str,
    event_version: int,
    producer: str,
    payload: TPayload,
    correlation_id: str | None = None,
    causation_id: str | None = None,
) -> EventEnvelope[TPayload]:
    """Mirrors packages/event-schemas' `createEnvelope()`. See its
    docstring for the "don't call twice for the same logical event"
    caveat around retries.
    """
    return EventEnvelope[TPayload](
        eventId=str(uuid.uuid4()),
        eventType=event_type,
        eventVersion=event_version,
        occurredAt=datetime.now(timezone.utc).isoformat(),
        correlationId=correlation_id or str(uuid.uuid4()),
        causationId=causation_id,
        producer=producer,
        payload=payload,
    )

File: packages/event-schemas/src/envelope.ts
import { randomUUID } from "node:crypto";

/**
 * Event envelope — TS mirror of schemas/event-envelope.schema.json, per
 * docs/architecture/Coding_Standards.md's Events section and
 * docs/adr/ADR-005-event-schema-versioning.md's versioning policy.
 *
 * There is no code generator wiring the JSON Schema, this type, and
 * apps/vision-service's Pydantic model together — they are kept in sync
 * by hand and verified by
 * apps/vision-service/tests/test_event_schema_sync.py (REQ-3.4), which
 * fails CI the moment `ENVELOPE_FIELD_NAMES` below drifts from either
 * the schema file's `properties` or the Pydantic model's fields.
 */
export interface EventEnvelope<TPayload = unknown> {
  readonly eventId: string;
  readonly eventType: string;
  readonly eventVersion: number;
  readonly occurredAt: string;
  readonly correlationId: string;
  readonly causationId: string | null;
  readonly producer: string;
  readonly payload: TPayload;
}

/**
 * Membership (not order) must match event-envelope.schema.json's
 * `properties` keys and the Pydantic `EventEnvelope`'s fields exactly —
 * see the sync check referenced above.
 */
export const ENVELOPE_FIELD_NAMES = [
  "eventId",
  "eventType",
  "eventVersion",
  "occurredAt",
  "correlationId",
  "causationId",
  "producer",
  "payload",
] as const;

export interface CreateEnvelopeInput<TPayload> {
  readonly eventType: string;
  readonly eventVersion: number;
  readonly producer: string;
  readonly payload: TPayload;
  /** Defaults to a fresh UUID — pass the originating HTTP request's correlation ID (REQ-3.11) when one exists. */
  readonly correlationId?: string;
  /** REQ-3.12: the eventId of the event that triggered this one. Omit/null for an event with no upstream event (e.g. produced directly from an HTTP request). */
  readonly causationId?: string | null;
}

/**
 * Builds a new envelope with a fresh `eventId`/`occurredAt`. Callers
 * that need the *same* `eventId` across retries (e.g.
 * apps/outbox-publisher re-publishing an outbox row) must generate the
 * id once and pass it through their own storage, not call this twice
 * for the same logical event.
 */
export function createEnvelope<TPayload>(
  input: CreateEnvelopeInput<TPayload>,
): EventEnvelope<TPayload> {
  return {
    eventId: randomUUID(),
    eventType: input.eventType,
    eventVersion: input.eventVersion,
    occurredAt: new Date().toISOString(),
    correlationId: input.correlationId ?? randomUUID(),
    causationId: input.causationId ?? null,
    producer: input.producer,
    payload: input.payload,
  };
}
